### PR TITLE
 episode level data 

### PR DESCRIPTION
**Pull request by daleyp24**
## JIRA Ticket
-  https://ihm-it.atlassian.net/browse/IHRACP-3057

## Description
- We should verify that our parsing library can parse episodes, and add the functionality if it does not exist. Identify what can be deprecated.

#### Major Changes:
Fields added:
Interactive 

Created uts for the following fields:
Episode_number
Season
Episode_Type 
Publish_Date
Image_url
Interactive 

Fields to remove (proposed): 
tv_rating
enddate
expiration
orignal air date
image -- seems to be null. url already exists  

Existing uts for episode fields:(line 61 of test)
description
author 
duration
explicit 
order 
subtitle
summary
enclosed url 
guid
linktitle -- just the title with no spaces
date time -- start date 

others:
last_modified -- populated internally through a condition.

## How To Test

1. pyenv exec tox. 
2. Verify all uts passed and the list above is accurate.
3. Verify all additional edge case listed are handled

Related PRs:
N/A

### CODE REVIEW

**Comment by KMichaelMcGaw** [comment]
We may want to edit the PR process for this repo, seems that PR reviews aren't necessary for merge, just the request

**Review by KMichaelMcGaw — APPROVED** [review]

**Review by DrewJoseyGit — APPROVED** [review]

**Review by IanNoyes — APPROVED** [review]

**Comment by daleyp24** [comment]
Cases we’ll need to account for when removing XSL Layer:

Episode DurationDuration Formatting Variations

Durations can be displayed in RSS feeds as either timestamp or seconds. 

      <itunes:duration>2785</itunes:duration> vs. <itunes:duration>50:19</itunes:duration>

<Content:encoded> vs <description> for episode description

<content:encoded> is an episode description that includes html formatting. It is our preference to use this field for our episode descriptions.

However, <content:encoded> is not a required field and some host services don’t provide it, so we need to be able to map to the <description> in those instances.

Format PubDates to reflect our internal Date formatting.

Here is an exampe of an RSS PubDate: <pubDate>Thu, 05 May 2022 22:11:03 +0000</pubDate>

We have logic in XSL to reformat PubDates to reflect our internal date/time formatting in the start_date and original_air_date fields in Postgres:

2022-05-25 22:11:03

Mapping <ihr:interactive> field

<is_interactive> is not not a standard RSS field. At this point, only one host service is providing this field specifically for us. 

We’ve added mapping in XSL to map the <ihr:interactive> to our <is_interactive> field.

Ensure <itunes:episodeType> is Lower Case

Some RSS feeds will pass the <itunes:episodeType> field as either upper or lower case. However, we only support lower case in our <item_type> field

Our XSL always sets the type data from <itunes:episodeType> as lower case.

Ensure <itunes:type> is Lower Case

Some RSS feeds will pass the <itunes:type> field as either upper or lower case. However, we only support lower case in our show level <type> field

Our XSL always sets the type data from <itunes:episodeType> as lower case.

File: pypodcastparser/Item.py [hunk 1]
```diff
@@ -103,11 +112,11 @@ def __init__(self, soup):
         self.set_dates_published()
 
     def set_time_published(self):
-        if self.published_date is None:
+        if self.published_date_string is None:
```
Comment (IanNoyes) [inline pypodcastparser/Item.py:116]: We're getting a large chunk of unit tests failing on some date formatting logic, wondering if it's tied to this change?  Can you elaborate on why you updated this section?  We might need to update the unit tests to account for this or find a different solution.

File: pypodcastparser/Podcast.py [hunk 2]
```diff
@@ -353,6 +358,10 @@ def set_published_date(self, tag):
         """Parses published date and set value"""
         try:
             self.published_date = tag.string
+            #Preserve the orignal tag for the start_date
+            self.published_date_string = tag.string
+            pubDate = datetime.datetime.strptime(self.published_date, "%a, %d %b %Y %H:%M:%S %Z")
+            self.published_date = datetime.datetime.strftime(pubDate,"%Y-%d-%m, %H:%M:%S")
```
Comment (IanNoyes) [inline pypodcastparser/Podcast.py]: This method is throwing a bunch of errors in our unit tests, think we need to take another look at what we're trying to accomplish here and see if there is a smother solution or if we need to update our Unit Tests.

**Review by IanNoyes — CHANGES_REQUESTED** [review]
Couple funky bits around the published_date.  We can take a look later and see if we can't resolve these bad boys.

File: pypodcastparser/Podcast.py (hunk 2 above)
Comment (daleyp24) [reply pypodcastparser/Podcast.py]: This logic was done to cover the following corner case: 
Here is an exampe of an RSS PubDate: Thu, 05 May 2022 22:11:03 +0000
We have logic in XSL to reformat PubDates to reflect our internal date/time formatting in the start_date and original_air_date fields in Postgres:
2022-05-25 22:11:03

PubDate is being used as start_date. The string pubDate is preserved for logic in the method on line 114 in Item and 154 in Podcasts

The unit tests are passing in my local env:
(pypod) Patrick-Daleys-MacBook-Pro:pyPodcastParser patrickdaley$ pyenv exec tox
py37 installed: attrs==21.4.0,beautifulsoup4==4.9.1,certifi==2020.6.20,chardet==3.0.4,coverage==6.4.1,datadiff==2.0.0,docopt==0.6.2,httmock==1.4.0,idna==2.10,importlib-metadata==1.7.0,iniconfig==1.1.1,lxml==4.5.2,packaging==21.3,pluggy==0.13.1,py==1.11.0,pyparsing==3.0.9,pytest==7.1.2,pytest-asyncio==0.18.3,requests==2.24.0,soupsieve==2.0.1,tomli==2.0.1,typing_extensions==4.2.0,urllib3==1.25.9,zipp==3.1.0
py37 run-test-pre: PYTHONHASHSEED='2168727786'
py37 run-test: commands[0] | python -m coverage run -m pytest --strict tests
==================================================================================================================== test session starts =====================================================================================================================
platform darwin -- Python 3.7.13, pytest-7.1.2, pluggy-0.13.1
cachedir: .tox/py37/.pytest_cache
rootdir: /Users/patrickdaley/pyPodcastParser
plugins: asyncio-0.18.3
asyncio: mode=legacy
collected 111 items                                                                                                                                                                                                                                          

tests/test_pyPodcastParser.py ..............................................................................................................                                                                                                           [ 99%]
tests/test_test.py .                                                                                                                                                                                                                                   [100%]

env: python 3.7.13

Comment (daleyp24) [reply pypodcastparser/Podcast.py]: @IanNoyes  
I am also curious about the implementation for the corner case below. I added a is_interactive tag in the tag field. Is that correct, or should this be at another layer?

Mapping ihr:interactive field
<is_interactive> is not not a standard RSS field. At this point, only one host service is providing this field specifically for us.
We’ve added mapping in XSL to map the ihr:interactive to our <is_interactive> field.

Comment (IanNoyes) [reply pypodcastparser/Podcast.py]: The place you put is_interactive is correct.  Good work!

Instead of trying to parse the dates by hand maybe look into using a library like [dateutil](https://github.com/dateutil/dateutil/tree/6b035517571e63b6a63a493740c5506ec1e5da44) to handle it.

Comment (IanNoyes) [reply pypodcastparser/Podcast.py]: I do see you included python-dateutil in the requirements file already, did you try this approach already?

Comment (daleyp24) [reply pypodcastparser/Podcast.py]: Yes, I attempted that approach. Ran into some issues and decided this approach was a better route.

**Review by KMichaelMcGaw — APPROVED** [review]

**Review by DrewJoseyGit — APPROVED** [review]
